metal/internal/datalist/filter.go

```diff
@@ -0,0 +1,245 @@
+package datalist
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+var (
+	matchByStringComparison = []string{"in", "re", "substring"}
+	matchByNumberComparison = []string{"less_than", "less_than_or_equal", "greater_than", "greater_than_or_equal"}
+)
+
+type commonFilter struct {
+	attribute string
+	values    []interface{}
+	all       bool
+	matchBy   string
+}
+
+func filterSchema(allowedAttributes []string) *schema.Schema {
+	return &schema.Schema{
+		Type: schema.TypeSet,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"attribute": {
+					Type:         schema.TypeString,
+					Description:  "The attribute used to filter. Filter attributes are case-sensitive",
+					Required:     true,
+					ValidateFunc: validation.StringInSlice(allowedAttributes, false),
+				},
+				"values": {
+					Type:        schema.TypeList,
+					Description: "The filter values. Filter values are case-sensitive. If you specify multiple values for a filter, the values are joined with an OR by default, and the request returns all results that match any of the specified values",
+					Required:    true,
+					Elem:        &schema.Schema{Type: schema.TypeString},
+				},
+				"all": {
```

Comment:
I contemplating suggesting that we drop `all` in favor of `match_by = "contains_all"`, but after looking at the implementation I see how `all` enables necessary boolean operations for other (all?) `match_by` operations.

Reply:
I find it useful the way it is, so you can combine also `all` with other match_by options like `re` (regexp) or `substring`.

```
# example to return all devices containing both "32-core" and "2.5Ghz" strings in the description
data "metal_plans" "example" {
    filter {
        attribute    = "description"
        values        = ["32-core", "2.5Ghz"]
        match_by  = "substring"
        all               = true
    }
}
```

I tried to add a good description to make it easy to understand how it works. If it's not, we can just remove "all" and same behavior can be achieved using several filters instead:

```
# same example to return all devices containing both "32-core" and "2.5Ghz" strings in the description
# using separated filteres over same attribute
data "metal_plans" "example" {
    filter {
        attribute    = "description"
        values        = ["32-core"]
        match_by  = "substring"
    }
    filter {
        attribute    = "description"
        values        = ["2.5Ghz"]
        match_by  = "substring"
    }
}